// @babylonjs/core/AudioV2/webAudio/webAudioBus.js
import { AudioBus } from "../abstractAudio/audioBus.js";
import { _StereoAudio } from "../abstractAudio/subProperties/stereoAudio.js";
import { _WebAudioBusAndSoundSubGraph } from "./subNodes/webAudioBusAndSoundSubGraph.js";
import { _SpatialWebAudio } from "./subProperties/spatialWebAudio.js";
/** @internal */
export class _WebAudioBus extends AudioBus {
    /** @internal */
    constructor(name, engine, options) {
        super(name, engine, options);
        this._spatial = null;
        this._stereo = null;
        this._subGraph = new _WebAudioBus._SubGraph(this);
        this.audioContext = engine.audioContext;
    }
    /** @internal */
    async init(options) {
        if (options.outBus) {
            this.outBus = options.outBus;
        }
        else {
            await this.engine.isReadyPromise;
            this.outBus = this.engine.defaultMainBus;
        }
        await this._subGraph.init(options);
        this.engine.addNode(this);
    }
    /** @internal */
    dispose() {
        super.dispose();
        this._spatial = null;
        this._stereo = null;
        this.engine.removeNode(this);
    }
    /** @internal */
    get inNode() {
        return this._subGraph.inNode;
    }
    /** @internal */
    get outNode() {
        return this._subGraph.outNode;
    }
    /** @internal */
    get spatial() {
        return this._spatial ?? (this._spatial = new _SpatialWebAudio(this._subGraph, this._spatialAutoUpdate));
    }
    /** @internal */
    get stereo() {
        return this._stereo ?? (this._stereo = new _StereoAudio(this._subGraph));
    }
    /** @internal */
    getClassName() {
        return "_WebAudioBus";
    }
}
_WebAudioBus._SubGraph = class extends _WebAudioBusAndSoundSubGraph {
    get _downstreamNodes() {
        return this._owner._downstreamNodes ?? null;
    }
    get _upstreamNodes() {
        return this._owner._upstreamNodes ?? null;
    }
};
